src/core_ocean/analysis_members/mpas_ocn_okubo_weiss.F

```diff
@@ -0,0 +1,1767 @@
+
+
+!***********************************************************************
+!
+!  routine ocn_aggregate_eddy_stats
+!
+!> \brief   Aggregate statistics of partial eddies
+!> \author  Andre Schmeisser
+!> \date    August 2014
+!> \details 
+!> Send all partial eddy statistics to IO node and merge them into
+!> statistics of complete eddies.
+!
+!-----------------------------------------------------------------------
+
+   subroutine ocn_aggregate_eddy_stats(dmInfo, use_lat_lon_coords, min_cells, &!{{{
+                                       numCCs, maxCCsPerDomain, &
+                                       origCCId, sumVol, numCells, &
+                                       wsPosX, wsPosY, wsPosZ, &
+                                       wsVelX, wsVelY, wsVelZ, &
+                                       OW_cc_id, nVertLevels, nCellsSolve)
+
+
+      !-----------------------------------------------------------------
+      !
+      ! input variables
+      !
+      !-----------------------------------------------------------------
+
+      type (dm_info), intent(in) :: dminfo
+      logical, intent(in) :: use_lat_lon_coords
+      integer, intent(in) :: numCCs, maxCCsPerDomain, min_cells
+      real (kind=RKIND), dimension(:), intent(in) :: origCCId, sumVol, numCells, &
+                             wsPosX, wsPosY, wsPosZ, wsVelX, wsVelY, wsVelZ
+      integer, intent(in) :: nVertLevels, nCellsSolve
+
+      !-----------------------------------------------------------------
+      !
+      ! input/output variables
+      !
+      !-----------------------------------------------------------------
+
+      integer, dimension(:,:), intent(inout) :: OW_cc_id
+
+      !-----------------------------------------------------------------
+      !
+      ! local variables
+      !
+      !-----------------------------------------------------------------
+
+      real (kind=RKIND), dimension(9 * maxCCsPerDomain) :: stats
+      real (kind=RKIND), dimension(9 * maxCCsPerDomain * dmInfo % nprocs) :: &
+         globalStats, aggregated
+
+      integer i, j, k, ierr, curId, listIdx, totalCCs, cullId
+
+      ! Multiplex all eddy statistics into one array so it can be send with
+      ! one MPI communication call
+      call mpas_timer_start("all gather", .false., CCAggGatherT)
+
+      stats = 0
+      do i = 1, numCCs
+         stats(9*(i-1) + 1) = origCCId(i)
+         stats(9*(i-1) + 2) = sumVol(i)
+         stats(9*(i-1) + 3) = wsPosX(i)
+         stats(9*(i-1) + 4) = wsPosY(i)
+         stats(9*(i-1) + 5) = wsPosZ(i)
+         stats(9*(i-1) + 6) = wsVelX(i)
+         stats(9*(i-1) + 7) = wsVelY(i)
+         stats(9*(i-1) + 8) = wsVelZ(i)
+         stats(9*(i-1) + 9) = numCells(i)
+      end do
+
+      ! Send all local statistics
+      call MPI_Allgather(stats, 9*maxCCsPerDomain, MPI_REALKIND, &
+                      globalStats, 9*maxCCsPerDomain, MPI_REALKIND, &
+                      dmInfo % comm, ierr)
```

Comment:
Not a problem. I didn’t really touch any of Andre’s code, but I can scan through and make sure he didn’t use any others. I believe he did use mpas_dmpar for the majority of the time.

-Jon

From: Doug Jacobsen <notifications@github.com<mailto:notifications@github.com>>
Reply-To: MPAS-Dev/MPAS-Release <reply@reply.github.com<mailto:reply@reply.github.com>>
Date: Wednesday, February 18, 2015 at 11:45 AM
To: MPAS-Dev/MPAS-Release <MPAS-Release@noreply.github.com<mailto:MPAS-Release@noreply.github.com>>
Cc: Jon Woodring <woodring@lanl.gov<mailto:woodring@lanl.gov>>
Subject: Re: [MPAS-Release] Okubo weiss 3.3 (#15)


In src/core_ocean/analysis_members/mpas_ocn_okubo_weiss.F<https://github.com/MPAS-Dev/MPAS-Release/pull/15#discussion_r24927899>:

> +      do i = 1, numCCs
> +         stats(9*(i-1) + 1) = origCCId(i)
> +         stats(9*(i-1) + 2) = sumVol(i)
> +         stats(9*(i-1) + 3) = wsPosX(i)
> +         stats(9*(i-1) + 4) = wsPosY(i)
> +         stats(9*(i-1) + 5) = wsPosZ(i)
> +         stats(9*(i-1) + 6) = wsVelX(i)
> +         stats(9*(i-1) + 7) = wsVelY(i)
> +         stats(9*(i-1) + 8) = wsVelZ(i)
> +         stats(9*(i-1) + 9) = numCells(i)
> +      end do
> +
> +      ! Send all local statistics
> +      call MPI_Allgather(stats, 9*maxCCsPerDomain, MPI_REALKIND, &
> +                      globalStats, 9*maxCCsPerDomain, MPI_REALKIND, &
> +                      dmInfo % comm, ierr)


This should use mpas_dmpar routines, instead of calling MPI_Allgather directly. At least, unless there's some reason we need to call MPAS_Allgather directly.

Generally, it's frowned upon to put direct MPI calls outside of mpas_dmpar.

—
Reply to this email directly or view it on GitHub<https://github.com/MPAS-Dev/MPAS-Release/pull/15/files#r24927899>.